resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeValidator.scala

```diff
@@ -0,0 +1,185 @@
+    POSSIBLE_RESOURCE_DEFINITIONS.foreach { rcp =>
+      val customResources: Map[String, String] = getCustomResourceValue(requestedResources, rcp)
+      val (standardResourceConfigKey: String, customResourceConfigKey: String) =
+        getResourceConfigKeys(rcp)
+
+      val errorMessage =
+        if (customResources.contains(customResourceConfigKey)) {
+          s"${rcp.resourceType} cannot be requested with config $customResourceConfigKey, " +
+              s"please use config $standardResourceConfigKey instead!"
+        } else {
+          ""
+        }
+      if (errorMessage.nonEmpty) {
+        printErrorMessageToBuffer(sb, errorMessage)
+      }
+    }
+
+    if (sb.nonEmpty) {
+      throw new SparkException(sb.toString())
+    }
+  }
+
+  /**
+   * Returns the requested map of custom resources,
+   * based on the ResourceConfigProperties argument.
+   * @return
+   */
+  private def getCustomResourceValue(
+      requestedResources: RequestedResources,
+      rcp: ResourceConfigProperties) = {
+    var customResources: Map[String, String] = null
+    (rcp.processType, rcp.runMode, rcp.resourceType) match {
+      case (ProcessType.executor, _, _) => customResources =
+          requestedResources.customExecutorResources
+      case (ProcessType.am, RunMode.client, _) => customResources =
+          requestedResources.customAMResources
+      case (ProcessType.driver, RunMode.cluster, _) => customResources =
+          requestedResources.customDriverResources
+    }
+    customResources
+  }
+
+  /**
+   * Returns a tuple of standard resource config key and custom resource config key, based on the
+   * processType and runMode fields of the ResourceConfigProperties argument.
+   * Standard resources are memory and cores.
+   *
+   * @param rcp
+   * @return
+   */
+  private def getResourceConfigKeys(rcp: ResourceConfigProperties): (String, String) = {
+    val standardResourceConfigKey: String = if (rcp.processType == ProcessType.am) {
+      s"spark.yarn.${rcp.processType}.${rcp.resourceType}"
+    } else {
+      s"spark.${rcp.processType}.${rcp.resourceType}"
+    }
+
+    var customResourceTypeConfigKey: String = ""
+    (rcp.processType, rcp.runMode) match {
+      case (ProcessType.am, RunMode.client) =>
+        customResourceTypeConfigKey += YARN_AM_RESOURCE_TYPES_PREFIX
+      case (ProcessType.driver, RunMode.cluster) =>
+        customResourceTypeConfigKey += YARN_DRIVER_RESOURCE_TYPES_PREFIX
+      case (ProcessType.executor, _) =>
+        customResourceTypeConfigKey += YARN_EXECUTOR_RESOURCE_TYPES_PREFIX
+    }
+
+    customResourceTypeConfigKey += rcp.resourceType
+
+    (standardResourceConfigKey, customResourceTypeConfigKey)
+  }
+
+  private[spark] def printErrorMessageToBuffer(sb: StringBuilder, str: String): Unit = {
+    sb.append(s"$ERROR_PREFIX$str\n")
+  }
+
+  private class ResourceConfigProperties(
+      val processType: ProcessType,
+      val runMode: RunMode = null,
```

Comment:
In my opinion, it's better not to use a default argument here. You construct only 6 instances of this anyway. Also consider making the type of runMode an Option[RunMode], so you can use None instead of null. It is my understanding that null is very unidiomatic in scala.

By the way,this is a good candidate for a scala case class, but it doesn't really matter.

Reply:
Modified runMode to be an Option and passed None instead of using default arguments.